Expose parsers for Longident.t

Octachron opened the pull request:
This PR proposes to expose the parsing functions for `Longident.t` defined in the main parser to the rest of the world.

Currently, some code in the compiler and in the wild (see ocaml-community/utop#301 or ocaml-ppx/ppxlib#44 ) assume that `Longident.parse` is able to properly parse `Longident.t` from strings.
This is unfortunately not the case: `Longident.parse` only implements a rough approximation of the syntactic rules used for `Longident`s in OCaml . Notably, the `Longident.parse` function fails spectacularly in presence of  functor application or dotted operators (e.g `(+.+.+)`).

To resolve this issue, this PR exposes a parsing functions for each major `Longident.t` subsets: values, constructors, modules, extended modules and (module) types in the `Parse.Longident` module.

Those new parser functions are then used to raise a syntax error when `-open` is provided invalid argument.

An unfortunate consequence of this change is that using the new parsing functions requires to pull the full parser as a dependency. This created a module name collision with the debugger parser and lexer module. This issue was fixed by prefixing the name of those modules.

trefis commented on boot/menhir/menhirLib.ml:
```diff
@@ -3513,5 +3513,5 @@ module MakeEngineTable (T : TableFormat.TABLES) = struct
 end
 module StaticVersion = struct
-let require_20181113 = ()
+let require_20190626 = ()
```
Is that version actually required? Or can we stick to an older version of menhir?

I guess the question I want to ask is: do we want to make sure we keep the oldest possible version there, or do we not care at all because it only affects compiler devs anyway?

gasche replied:
I would hesitate between "we don't care" and "we should only accept increases to this field to avoid back-and-forth; it's easy for users of older versions to upgrade anyway".

rgrinberg commented:
IIRC, in ppxlib (and ppx) it was [stated](https://github.com/ocaml-ppx/ppxlib/pull/98) that we wouldn't rely on the compiler-libs api for this job even if such a function was provided. It's a bit of a shame that code has to be duplicated, but an unstable API is not an appropriate dependency for key libraries such as ppxlib.

gasche commented:
This PR is spot on for solving the Merlin issue. Sorry for missing it earlier. I'll do a review.

gasche commented on .gitignore:
```diff
@@ -72,9 +72,9 @@ _build
-/debugger/lexer.ml
-/debugger/parser.ml
-/debugger/parser.mli
+/debugger/debugger_lexer.ml
```
I would prefer if the `Lexer => Debugger_lexer` change was made in a first, separate commit.

Octachron replied:
I will split the commit once the timeline is stabilized.

Octachron replied:
History has been rewritten and the commit split.

gasche commented on parsing/longident.mli:
```diff
@@ -29,6 +29,9 @@ val flatten: t -> string list
 val parse: string -> t
+(** This function may return syntactically invalid {!t}.
+    If you need precise parsing function , consider using one of the functions
+    in {!Parse.Longident}  *)
```
I would suggest changing the name of this function to `parse_flat` or some other name that highlights the limitation. Otherwise it's just too tempting to use it. The very point of this PR is that downstream users should *not* use this function, so forcing them to think about what to use instead would be nice.

Octachron replied:
It sounds like a good idea.

gasche commented on parsing/parse.mli:
```diff
@@ -27,3 +27,68 @@ val use_file : Lexing.lexbuf -> Parsetree.toplevel_phrase list
+      All variants parse slightly different sets of {!Longident.t}.
+  *)
+
+  val value: ?cmd:string -> string -> (Longident.t,exn) result
```
I'm not fond of the API.

I think that we should expose `Lexing.lexbuf -> Longident.t` functions, as above. In particular, Merlin does propagate location information up to the point where they want to parse longidents, so they would be able to initialize the lexbuf location in interesting ways.

We could provide a generic wrapper in `parse`, for example `(lexbuf -> 'a) -> ?cmd:string -> string -> ('a, exn) result`. I'm not convinced this is necessary (and it is doing many orthogonal things at once: taking string inputs, changing error messages, and reifying exceptions into the result monad).

Octachron replied:
Indeed, `Lexing.lexbuf -> Longident.t` seems better, even for the compiler.

gasche commented on parsing/parser.mly:
```diff
@@ -1119,6 +1129,31 @@ parse_pattern:
     { $1 }
 ;
 
+parse_mty_longident:
```
The convention I follow in my own parsers is `mty_longident_eof` for these starting actions, I personally find it clearer than yours. (The existing code follows no convention anyway.)

Octachron replied:
I was trying to respect the convention used in parse_{core_type,expression,pattern}, but I have no strong opinion on the matter.

gasche replied:
Ah, I hadn't noticed that this convention is (half-)followed in the existing codebase. It's not a very good convention, but also fairly orthogonal to the present PR; it's fine to leave it as-is.

gasche reviewed (commented):
I like the idea (I independently came to the conclusion that it would be the right approach to avoid `Longident.parse`), but I think that the API is too specific to the existing use-cases in the compiler, and that the standard API would be better (possibly paired with helper functions to provide simpler entry points).

gasche commented:
In https://github.com/ocaml/merlin/issues/949#issuecomment-568299470, @Octachron asks:

> Should I add a generic Longident.t parser to ocaml/ocaml#9021?

Yes, I believe this would be nice for the cases when we don't know the syntactic category. The parser should accept any of the inputs accepted by the other parsers, so I think that it should:

1. Accept extended module paths (with `Lapply` node)
2. Accept infix/literal final lidents

I believe that this requires a bit of refactoring of the grammar, but it should not be difficult. (You need to factor the infix/literal part of `val_longident` out of the current non-extended-longident choice.)

gasche commented on parsing/parse.mli:
```diff
@@ -29,14 +29,28 @@ val expression : Lexing.lexbuf -> Parsetree.expression
+     compiler. Contrarily to all specialized parsers below,
+     [any] does not enforce that a module path that contains
+     an application ends on a valid identifier for either a type
+     or a module type. For instance, [F(X).(::)] is not accepted by OCaml parser
```
Here is an explanation that I would find clearer:

```
In valid OCaml syntax, only value-level identifiers may end with infix operators [Foo.( + )].
On the other hand, in value-level identifiers the module path [Foo] must be simple
([M.N] rather than [F(X)]): functor applications may only appear in type-level identifiers.
As a consequence, a path such as [F(X).( + )] is not a valid OCaml identifier,
but it is accepted by this function.
```

(Feature request: infix module type operators.)

Octachron replied:
The explanation is better, but I think it warrants updating the first sentence to
```
     However, this function [any] accepts some inputs which are not accepted by the
     compiler, because they combine functor applications and infix operators.
     In valid OCaml syntax, only value-level identifiers may end with infix
     operators [Foo.( + )].
     Moreover, in value-level identifiers the module path [Foo] must be simple
     ([M.N] rather than [F(X)]): functor applications may only appear in type-level identifiers.
     As a consequence, a path such as [F(X).( + )] is not a valid OCaml identifier;
     but it is accepted by this function.
```

gasche reviewed (commented):
It's probably not your fault, but I find the resulting parsing rules fairly confusing. `val_extra_ident` denotes identifiers that are valid for values (but not the other classes), but can be prefixed by a module path. `constr_extra_ident` denotes identifiers that are valid for constructors (but not the other classes), but *cannot* be prefixed by a module path.

Ideally I would imagine something like the following:
- a parametrized rule `longident(path, tip)` parametrized over grammars for prefix paths (extended or not?) and tip / lastname (infixes or not?)
- in the rare cases that have it (only `constr_longident`?), an extra subcategory with a name that sticks out for non-prefixable tips.

I'm not sure if this can be made conflict-avoiding in the same way as the (fairly hard to understand) `%prec below_DOT` stuff.

Octachron commented:
In the current code `constr_extra_indent` contains `::`, so the corresponding identifiers may or may not be prefixed by a module path. I am not sure if it more or less confusing that way.

I am also wondering if I should take the time to make the generic `Longident.parser` match exactly the class of accepted `Longident.t` or if the current approximation is fine?

gasche commented:
I think that the current approximation is fine. Note that if you wanted to capture exactly the valid longidents, the easiest way would be to work outside the grammar, by (copying the lexbuf state and) calling each class-specific parser in turn.

Octachron commented:
I reworked a bit the constructor identifier classes to make the non-prefixable class more explicit. I also added a `mk_longident` functions in order to factorizes the common definition for the various longident classes. There is still a quite distinct  `constr_longident` definition, but this seems necessary to avoid conflicts between `(A.B) . x` and `(A) . B.x`.

gasche commented on testsuite/tests/compiler-libs/test_longident.ml:
```diff
@@ -56,32 +58,128 @@ let last_dot_apply = L.last
+let test specialized s =
+  { flat = L.parse_flat s;
+    spec = specialized (Lexing.from_string s);
+    any_ = PL.any (Lexing.from_string s)
```
If we know that `spec` and `any_` should always be the same, I'm not convincing by showing them both in the expect-result and eyeballing that they are the same. I would rather have a field `any_correct = (PL.any ... = spec)` for which I could simply check that it is always `true` in the expected result.

Octachron replied:
Done.

gasche commented on toplevel/genprintval.ml:
```diff
@@ -547,10 +547,14 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
            from its name *)
+        let lid = match Parse.Longident.constr (Lexing.from_string name) with
+          | exception Syntaxerr.Error _ ->
+              (* e.g. constructors in non-instancied functor: F(X).A *)
```
instantiated?

gasche reviewed (commented):
I like the new `parser.mly` changes, and I am generally ok with the PR as-is. I haven't approved yet because there is a lot of history back-and-forth and `.depend` noises and menhir-version noise, etc. I would rather do a clear review again when you have cleaned up everything you intend to clean up.

Octachron committed:
debugger: rename parser and lexer modules

Octachron committed:
compilerlibs: expose parsers for 'Longident.t's

gasche commented on parsing/longident.mli:
```diff
@@ -28,7 +28,10 @@ type t =
+val parse_flat: string -> t
+(** This function may return syntactically invalid {!t}.
+    If you need precise parsing function, consider using one of the functions
+    in {!Parse.Longident}  *)
```
Sorry for asking only now, but do we actually need this function at all anymore? If we agree that changing the name (and thus breaking experimental-api-user code) is the the right choice, why don't we just get rid of it completely?

gasche replied:
The recommend replacement would be something like

```ocaml
   Parse.Longident.any (Lexing.from_string input)
```

or using a more precise syntactic category.

gasche replied:
Another choice would be to only mark `Location.parse` deprecated, instead of removing it altogether. The documentation/deprecation comment should explain clearly why the function is broken.

cc @diml, as our ocaml-migrate-parsetree spokeperson: is something terribly bad going to happen if we remove `Longident.parse` in favor of actually-correct parser entry points?

diml replied:
There are a few consequences:

- if you remove the function entirely, that's going to cause quite a bit of churn in omp and a few other projects. Longident is currently not snapshotted by omp
- if you replace it by calling the lexer and parser, that's going it to make it slower and allocate more. This function tend to be called a lot in ppx rewriters. For instance, in ppxlib we have a function `evar : loc:Location.t -> string -> expression` that creates a variable expression from a string. This function calls `Longident.parse` and is used quite a bit. I have seen similar things in other projects. This could have an impact on the performance of ppx rewriters, or could be lost in the noise. It's difficult to say without testing

Personally, I'd just keep the current `Longident.parse` function for now. In the upcoming `ppx` library, we could stop relying on `Longident.parse` and then we could delete `Longident.parse` in a couple of years.

Octachron replied:
Keeping `parse` sounds fine in this case (wouldn't it be nice if we had some kind of alerts telling users that a function is expected to be removed?). For ppxlib, wouldn't it work to expose an evar function with type: `loc:Location.t -> string list -> expression`? This function would avoid half of the work in `Longident.parse`, cover the same subclass of `Longident.t`s, and avoid all surprising error cases.

gasche replied:
In fact, a `string list -> Longident.t` function already exists as `Longident.unflatten`. After thinking about the issue, I think that ppx authors should use `unflatten` instead (they can do the splitting on their side if desired, at least the semantics is clear to the code author), and I remain of the opinion that we should remove `Longident.parse`.

gasche replied:
(In particular, ppx authors switching to `unflatten` will have code that is backward-compatible, so transition pains should be limited.)

diml replied:
I agree with you both, and the truth is maybe our `evar` function should do no parsing at all and consider that the argument is not a long identifier but just a variable name. We'll certainly think about this for the upcoming ppx project.

But right now, I'm personally not excited by the amount of downstream work this represent and I'd prefer not to have to deal with it now.

That said, we have been in this situation of few years now, so even if removing `Longident.parse` will cause some churn, it's nothing we haven't done before.

To give a shorter and clearer answer to "will something terribly bad happen if we remove Longident.parse": nothing terribly bad will happen and we would be able to cope with the change. However, given that `Longident` is not currently snapshotted by omp, this won't be straightforward. Instead, when upgrading omp we are going to need to turn our brains and think about how to handle the removal of `Longident.parse`. The solution could be quite annoying to implement.

gasche replied:
The present PR was revived by the fact that the use of `Longident.parse` by downstream users creates bug (in the specific example discussed it was a bug in Merlin). They rely on it, but a good fraction of those use-cases are wrong (probably the large majority of them), and it's not obvious to notice that they are wrong (it only occurs on inputs that are less frequent in the wild, eg. `( +. )` instead of `foo`). So yes removing the function would cause transition pain, but it will also force users to fix bugs that they are not aware of, so it is not just a waste a time.

Re. snapshotting Longident in ocaml-migrate-parsetree: for this particular change, we would just be removing one function. Would it be so bad if the module remained non-snapshotted? (non-snapshot?) Without snapshotting, the issue is that (probably wrong) code users write will fail to compile on recent OCaml versions, and they have to use a different function which exists on all supported version; that doesn't sound so bad to me (it's not as good as the usual m-o-p guarantees, but those only covers snapshotted modules indeed).

diml replied:
IIUC, right now on invalid input `Longident.parse` doesn't fail and instead produces an invalid `Longident.t`? If yes, what about making it fail on invalid input? The error doesn't have to be fancy, it can just be `invalid_argument` and the caller has to use the proper parser if they want a proper error. That seems like the simplest solution to me.

> Re. snapshotting Longident in ocaml-migrate-parsetree: for this particular change, we would just be removing one function. Would it be so bad if the module remained non-snapshotted? (non-snapshot?) Without snapshotting, the issue is that (probably wrong) code users write will fail to compile on recent OCaml versions, and they have to use a different function which exists on all supported version; that doesn't sound so bad to me (it's not as good as the usual m-o-p guarantees, but those only covers snapshotted modules indeed).

￼That'd be fine yes, but still quite a lot of code to update.

Octachron replied:
It is worse than that: `Longident.parse` can produce invalid outputs from valid inputs. And indeed, it never fails. I am not sure if it makes sense to raise an exception for corner cases that people *keep* forgetting.

gasche replied:
The result of returning an exception is that users would see preprocessing-time crashes (or merlin-query-time crashes) instead of silent bad-code generation. I think that asking extension authors to fix their code to use correct functions instead is better for the ecosystem.

diml replied:
could you deprecate or put an alert on this function rather than delete it immediately?

Octachron replied:
Deprecating the function sounds fine to me, but wouldn't it create essentially the same issue as deleting it for all dune users?

diml replied:
No. Dune users will indeed get an error during development, but released packages won't start failing to build because of the deprecation. So if you have a ppx that you are not actively developing, you are free to ignore the deprecation initially.

Octachron replied:
Ok, this sounds like an improvement. I made the change (and kept the `parse` name for now).

diml replied:
Thanks!

gasche replied:
I'm still grumbling at the idea of keeping a function that is known to be very broken (it fails on `( +. )`, not just on extended module paths), but let's consider this a good compromise for now. I think the deprecation message should be improved. Current version:

```
    This function may parse incorrectly valid {!t} and return syntactically
    invalid {!t}s.
    If you need precise parsing function, consider using {!Parse.longident} or
    one of the specialized functions in {!Parse}.
```

The deprecation message should (1) list cases that are known to be broken, and (2) mention `unflatten` with a clear enough description of its specification that users know when to adopt it instead of calling `Parse`. In particular, ocaml-migrate-parsetree users do not have an easy access to the `Parse` module, but they can easily change to another function within `Longident`.

Proposal:

```
This function is broken on identifiers that are not just "Word.Word.word";
for example, it returns incorrect results on infix operators
and extended module paths.

If you know that the identifiers you parse are of this simple form
(lists of dot-separated identifiers), you should use {!unflatten} instead
which is safer and faster. {!unflatten} is available since OCaml 4.06.0.

If you want to parse any identifier correctly, use the long-identifiers
functions from the {!Parse} module, in particular {!Parse.longident}.
They are available since OCaml 4.11, and also provide proper
input-location support.
```

Octachron replied:
Being more explicit on the issue at hand is clearly better.
I am not so fond of the second paragraph because the use case of `unflatten` is quite different. Maybe something like:
```
   If you want to generate long identifiers that are a list of
   dot-separated identifiers, the function {!unflatten} is safer and faster.
   {!unflatten} is available since OCaml 4.06.0.
```
in order to emphasis the contrast between `Parse.longident` (for parsing (unsafe) user inputs), and `unflatten` (for generating the most common class of longidents)?

gasche replied:
Yes, I like your second paragraph better.

gasche reviewed (commented):
The PR is clean but there is a last-minute exciting turn of events, see the inline comments.

gasche commented on parsing/parse.ml:
```diff
@@ -121,7 +121,14 @@ and core_type = wrap_menhir Parser.Incremental.parse_core_type
+  let mod_ext = wrap_menhir Parser.Incremental.parse_mod_ext_longident
+  let simple_module = wrap_menhir Parser.Incremental.parse_mod_longident
+  let typ = wrap_menhir Parser.Incremental.parse_mty_longident
+end
```
Nitpick: I'm not fond of `mod_ext` and `simple_module`. The mod/module discrepancy is not nice, and I would like to see `path` in the name, because I think of a simple module *path*, not a simple *module*. 

If it was me, I would do away with the Longident module and expose

```ocaml
val_ident
constr_ident
typ_ident
simple_module_path
{ext,full}_module_path (* extended_ is too long *)
```

Octachron replied:
The name "extended module path" is already used in the manual (to describe which kind of modules are accepted in open in signature).

Octachron replied:
I applied the change, mainly for the `Longident.any` ⇒ `longident` renaming.

gasche reviewed (approved):
The whole thing seems good to me now. (I assume you want to rebase your commits before a merge.)

SkySkimmer commented on Changes:
```diff
@@ -111,6 +111,9 @@ Working version
 - #9060: ensure that Misc.protect_refs preserves backtraces
   (Gabriel Scherer, review by Guillaume Munch-Maccagnoni and David Allsopp)
 
+- #9061: expose compiler Longident.t parsers
```
Wrong number

gasche replied:
Thanks! Fixed in 09da7dd659b2c3c685cd1bd6f42aba9e87e42036.
